File: AI_LAB2/EX7/Nhom2_28_B2113306_KimKhanhDang_Ex7/Nhom2_28_B2113306_KimKhanhDang_BFS_Heuristic.cpp
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<algorithm>
#include<vector>
#define ROWS 3
#define COLS 3
#define empty_p 0
#define max_operator 4
#define maxlength 500

using namespace std;

const char* action[] = {
	"First state",
	"Move cell Empty to Up",
	"Move cell Empty to Down",
	"Move cell Empty to Left",
	"Move cell Empty to Right"
};

typedef struct{
	int puzzle[ROWS][COLS];
	int emptyRow;
	int emptyCol;
}State;

void print_State(State state){
	int row, col;
	printf("\n----------\n");
	for(row=0; row<ROWS; row++){
		for(col=0; col<COLS; col++)
			printf("|%d ",state.puzzle[row][col]);
		printf("|\n");
	}
	printf("----------\n");
}


int compareState(State state1, State state2){
	if(state1.emptyCol != state2.emptyCol || state1.emptyRow != state2.emptyRow)
		return 0;
	int row, col;
	for(row=0; row<ROWS; row++)
		for(col=0; col<COLS; col++)
			if(state1.puzzle[row][col] != state2.puzzle[row][col])
				return 0;
	return 1;		
}

int goalCheck(State state, State goal){
	return compareState(state, goal);
}

int upOperator(State state, State *result){
	*result = state;
	int empRowCurrent = state.emptyRow, empColCurrent = state.emptyCol;
	if(empRowCurrent > 0){
		result->emptyRow = empRowCurrent - 1;
		result->emptyCol = empColCurrent;
		result->puzzle[empRowCurrent][empColCurrent] = state.puzzle[empRowCurrent-1][empColCurrent];
		result->puzzle[empRowCurrent-1][empColCurrent] = empty_p;
		return 1;
	}
	return 0;
}

int downOperator(State state, State *result){
	*result = state;
	int empRowCurrent = state.emptyRow, empColCurrent = state.emptyCol;
	if(empRowCurrent < 2){
		result->emptyRow = empRowCurrent + 1;
		result->emptyCol = empColCurrent;
		result->puzzle[empRowCurrent][empColCurrent] = state.puzzle[empRowCurrent+1][empColCurrent];
		result->puzzle[empRowCurrent+1][empColCurrent] = empty_p;
		return 1;
	}
	return 0;
}

int leftOperator(State state, State *result){
	*result = state;
	int empRowCurrent = state.emptyRow, empColCurrent = state.emptyCol;
	if(empColCurrent > 0){
		result->emptyRow = empRowCurrent;
		result->emptyCol = empColCurrent - 1;
		result->puzzle[empRowCurrent][empColCurrent] = state.puzzle[empRowCurrent][empColCurrent-1];
		result->puzzle[empRowCurrent][empColCurrent-1] = empty_p;
		return 1;
	}
	return 0;
}

int rightOperator(State state, State *result){
	*result = state;
		int empRowCurrent = state.emptyRow, empColCurrent = state.emptyCol;
	if(empColCurrent < 2){
		result->emptyRow = empRowCurrent;
		result->emptyCol = empColCurrent + 1;
		result->puzzle[empRowCurrent][empColCurrent] = state.puzzle[empRowCurrent][empColCurrent+1];
		result->puzzle[empRowCurrent][empColCurrent+1] = empty_p;
		return 1;
	}
	return 0; 
}


int callOperators(State state, State *result, int opt){
	switch(opt){
		case 1: return upOperator(state, result);
		case 2: return downOperator(state, result);
		case 3: return leftOperator(state, result);
		case 4: return rightOperator(state, result);
		default: 
			printf("Cannot call operators");
			return 0;
	}
}

int heuristic1(State state, State goal){
	int row, col;
	int count = 0;
	for(row=0; row<ROWS; row++)
		for(col=0; col<COLS; col++){
			if(state.puzzle[row][col] != goal.puzzle[row][col])
				count++;
		}
	return count;
}

int heuristic2(State state, State goal){
	int cnt = 0;
	int row, col, row_g, col_g;
	for(row=0; row<ROWS; row++)
		for(col=0; col<COLS; col++)
			if(state.puzzle[row][col] != empty_p)
				for(row_g=0; row_g<ROWS; row_g++)
					for(col_g=0; col_g<COLS; col_g++)
						if(state.puzzle[row][col] == goal.puzzle[row_g][col_g]){
							cnt+=abs(row-row_g) + abs(col-col_g);
							row_g=ROWS;
							col_g=COLS;
						}
	return cnt;
}

typedef struct Node{
	State state;
	struct Node* parent;
	int no_func;
	int heuristic;
}Node;

//Node* findState(State state, vector<Node*>& v, vector<Node*>::iterator& p){
//	vector<Node*>::iterator it = v.begin();
//	if(v.size() == 0) return NULL;
//	while(it != v.end()){
//		if(compareState((*it)->state, state)){
//			p = it;
//			return *it;
//		}
//		it = v.erase(it);
//	}
//	return NULL;	
//}

Node* findState(State state, vector<Node*>& v, vector<Node*>::iterator& p){
    for (p = v.begin(); p != v.end(); ++p) {
        if (compareState((*p)->state, state)) {
            return *p;
        }
    }
    return NULL;
}


bool compareHeuristic(Node* a, Node* b){
	return a->heuristic > b->heuristic;
}

Node* Best_First_Search(State state, State goal){
	vector<Node*> OPEN_BFS;
	vector<Node*> CLOSE_BFS;	
	Node* root = new Node();
	root->state = state;
	root->parent = NULL;
	root->no_func = 0;
	root->heuristic = heuristic2(root->state, goal);
	OPEN_BFS.push_back(root);
	while(!OPEN_BFS.empty()){
		Node* node = OPEN_BFS.back();
		OPEN_BFS.pop_back();
		if(goalCheck(node->state, goal))
			return node;
		CLOSE_BFS.push_back(node);
		int opt;
		for(opt=1; opt<=max_operator; opt++){
			State newstate;
			newstate = node->state;
			if(callOperators(node->state, &newstate, opt)){
				Node* newNode = new Node();
				newNode->state = newstate;
				newNode->parent = node;
				newNode->no_func = opt;
				newNode->heuristic = heuristic2(newstate, goal);
				vector<Node*>::iterator pos_Open, pos_Close;
				Node* nodeFoundOpen = findState(newstate, OPEN_BFS, pos_Open);
				Node* nodeFoundClose = findState(newstate, CLOSE_BFS, pos_Close);
				if (nodeFoundOpen == NULL && nodeFoundClose == NULL){
					OPEN_BFS.push_back(newNode);
				}else if(nodeFoundOpen != NULL && nodeFoundOpen->heuristic > newNode->heuristic){
					OPEN_BFS.erase(pos_Open);
					OPEN_BFS.push_back(newNode);
				}else if(nodeFoundClose != NULL && nodeFoundClose->heuristic > newNode->heuristic){
					CLOSE_BFS.erase(pos_Close);
					OPEN_BFS.push_back(newNode);
				}
			}
		}
		sort(OPEN_BFS.begin(), OPEN_BFS.end(), compareHeuristic);
	}
	return NULL;
}

void print_Way(Node* node){
	vector<Node*> vectorPrint;
//	listPrint.clear();
	while(node->parent != NULL){
		vectorPrint.push_back(node);
		node = node->parent;
	}
	vectorPrint.push_back(node);
	int no_act = 0;
	for(int i = vectorPrint.size() - 1; i >= 0; i--){
		printf("\nAction %d: %s", no_act, action[vectorPrint.at(i)->no_func]);
		print_State(vectorPrint.at(i)->state);
		no_act++;
	}
}

int main(){
	State state, goal;
	state.emptyRow = 1;
	state.emptyCol = 1;
	state.puzzle[0][0] = 3;
	state.puzzle[0][1] = 4;
	state.puzzle[0][2] = 5;
	state.puzzle[1][0] = 1;
	state.puzzle[1][1] = 0;
	state.puzzle[1][2] = 2;
	state.puzzle[2][0] = 6;
	state.puzzle[2][1] = 7;
	state.puzzle[2][2] = 8;
	
	goal.emptyRow=0;
	goal.emptyCol=0;
	goal.puzzle[0][0] = 0;
	goal.puzzle[0][1] = 1;
	goal.puzzle[0][2] = 2;
	goal.puzzle[1][0] = 3;
	goal.puzzle[1][1] = 4;
	goal.puzzle[1][2] = 5;
	goal.puzzle[2][0] = 6;
	goal.puzzle[2][1] = 7;
	goal.puzzle[2][2] = 8;
	
	Node* p = Best_First_Search(state, goal);
    if (p != NULL) {
        print_Way(p);

        while (p != NULL) {
            Node* temp = p;
            p = p->parent;
            delete temp;
        }
    } else {
        printf("No solution found.\n");
    }
	
	return 0;
}

File: AI_LAB2/EX7/Nhom2_28_B2113306_KimKhanhDang_BFS_Heuristic.py
ROW = 3
COL = 3
GOAL = ((0, 1, 2), (3, 4, 5), (6, 7, 8))
FIRST_STATE = [[3, 4, 5], [1, 0, 2], [6, 7, 8]]
dr = (0, 1, 0, -1)
dc = (1, 0, -1, 0)

action = [
    "move cell empty to Right", "move cell empty to Down",
    "move cell empty to Left", "move cell empty to Up",
    "first state"
]


class State:
    def __init__(self, data = FIRST_STATE, row=None, col=None):
        self.data = [list(row) for row in data]
        self.row = row
        self.col = col

    def as_tuple(self):
        return tuple(tuple(row) for row in self.data)

    def goal_check(self):
        return tuple(tuple(row) for row in self.data) == GOAL

    def update_position_cell_empty(self):
        for r, row in enumerate(self.data):
            for c, col in enumerate(row):
                if col == 0:
                    self.row = r
                    self.col = c


def print_state(state):
    for row in state.data:
        print(row)


def heuristic(state):
    cost = 0
    for r in range(ROW):
        for c in range(COL):
            value = state.data[r][c]
            goal_row, goal_col = divmod(value, COL)
            cost += abs(r - goal_row) + abs(c - goal_col)
    return cost


def move(state, r, c):
    row, col = state.row, state.col
    rnew, cnew = row + r, col + c
    if 0 <= rnew < ROW and 0 <= cnew < COL:
        data = [list(row) for row in state.data]
        data[row][col], data[rnew][cnew] = data[rnew][cnew], data[row][col]
        new_state = State(data, rnew, cnew)
        return new_state
    return None


class Node:
    def __init__(self, state, parent, noFunction, cost, id):
        self.state = state
        self.parent = parent
        self.noFunction = noFunction
        self.cost = cost
        self.id = id

    def priority(self):
        return heuristic(self.state)

    def __lt__(self, other):
        return self.priority() < other.priority() or (self.priority() == other.priority() and self.id < other.id)

def best_first_search(state):
    open_set, close_set = [Node(state, None, -1, 0, 0)], set()
    close_set.add(state.as_tuple())
    id_counter = 1
    while open_set:
        open_set.sort(key=lambda node: node.priority())
        node = open_set.pop(0)
        if node.state.goal_check():
            return node
        for i, (r, c) in enumerate(zip(dr, dc)):
            if node.noFunction == (i + 2) % 4:
                continue
            new_state = move(node.state, r, c)
            if new_state is None:
                continue
            data_tuple = new_state.as_tuple()
            if data_tuple not in close_set:
                open_set.append(Node(new_state, node, i, node.cost + 1, id_counter))
                close_set.add(data_tuple)
                id_counter += 1
    return None


def print_path(node):
    path = []
    while node:
        path.append(node)
        node = node.parent
    path.reverse()
    for i, item in enumerate(path):
        print("\nAction {}: {}".format(i, action[item.noFunction]))
        print_state(item.state)


if __name__ == "__main__":
    state = State()
    state.update_position_cell_empty()
    solution_node = best_first_search(state)
    if solution_node:
        print_path(solution_node)
    else:
        print("No solution found.")
